ArjunaJTA/jta/classes/com/arjuna/ats/internal/jta/transaction/arjunacore/jca/XATerminatorImple.java

```diff
@@ -241,6 +241,12 @@ public int prepare (Xid xid) throws XAException
 
             switch (tx.doPrepare())
             {
+            case TwoPhaseOutcome.PREPARE_ONE_PHASE_COMMITTED:
+                // Should never happen Would be great if there was heuristic commit :(
+                SubordinationManager.getTransactionImporter()
+                        .removeImportedTransaction(xid);
+                jtaLogger.i18NLogger.fatalSubordinate1PCDuringPrepare(xid);
+                throw new XAException(XAException.XAER_RMERR);
```

Comment:
OK so the normal behaviour for this optimisation is to commit the last participant immediately (if the others all returned READONLY) but that such behaviour is wrong for subordinates.

So you throw this exception is in case you failed to disable the optimisation for subordinates, ie is it just defensive programming in case the logic that disables the optimisation breaks?